crates/iota-core/src/starfish_adapter.rs

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) 2025 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{sync::Arc, time::Duration};
+
+use arc_swap::{ArcSwapOption, Guard};
+use iota_types::{
+    error::{IotaError, IotaResult},
+    messages_consensus::{ConsensusTransaction, ConsensusTransactionKind},
+};
+use starfish_core::{ClientError, TransactionClient};
+use tap::prelude::*;
+use tokio::time::{Instant, sleep};
+use tracing::{error, info, warn};
+
+use crate::{
+    authority::authority_per_epoch_store::AuthorityPerEpochStore,
+    consensus_adapter,
+    consensus_adapter::{BlockStatusReceiver, ConsensusClient},
+    consensus_handler::SequencedConsensusTransactionKey,
+};
+
+/// Gets a client to submit transactions to Starfish, or waits for one to be
+/// available. This hides the complexities of async consensus initialization and
+/// submitting to different instances of consensus across epochs.
+#[derive(Default, Clone)]
+pub struct LazyStarfishClient {
+    client: Arc<ArcSwapOption<TransactionClient>>,
+}
+
+impl LazyStarfishClient {
+    pub fn new() -> Self {
+        Self {
+            client: Arc::new(ArcSwapOption::empty()),
+        }
+    }
+
+    async fn get(&self) -> Guard<Option<Arc<TransactionClient>>> {
+        let client = self.client.load();
+        if client.is_some() {
+            return client;
+        }
+
+        // Consensus client is initialized after validators or epoch starts, and cleared
+        // after an epoch ends. But calls to get() can happen during validator
+        // startup or epoch change, before consensus finished initializations.
+        // TODO: maybe listen to updates from consensus manager instead of polling.
```

Comment:
Maybe Tokio's watch + using `wait_for` is a good idea too https://docs.rs/tokio/latest/tokio/sync/watch/struct.Receiver.html#method.wait_for :)